### DISCUSSION TITLE
'use server' within server components

### DISCUSSION THREAD

**Discussion by Nicky-Sa** [vercel/next.js#68505] [App Router]
Hello everyone. I have a question about Next.js "use server" directive.
I know that we can have server actions in 2 ways: 
1. In a separate file marked with 'use server'
2. [inside of a server component, at the top of a function](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#server-components).

What I don't understand is in what situation we might need to use the second way? If we're in a component, either we want to get or post data. In server components, if we're getting data, we just use the plain [fetch with async/await](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#server-components) pattern, no server action needed. Also, in server components, we can't have user interaction, so we can't post data based on user input. Again, no need for server action!
Then when can we declare a function to run on server? Isn't the component already running on server?

**Comment by sofly** [comment]
@Nicky-Sa Hello. I think you can use the second approach when you don't want to move the server function to another module and when you don't want to import the server function directly from a client component.

```tsx
import { Button } from './button';
import { Link } from './link';

interface Props {
  searchParams: {
    filter: string;
  };
}

export default function Page({ searchParams: { filter } }: Props) {
  async function doSomeAction() {
    'use server';

    console.log('do something');
  }

  if (filter) {
    return <Link onClick={doSomeAction}>Click me</Link>;
  }

  return <Button onClick={doSomeAction}>Click me</Button>;
}
```

It's nice that you don't have to create a wrapper for the link and button every time you need to import a server action when you want to pass server actions to client components. Does it make sense?

**Reply by Nicky-Sa** [reply]
I don't think this pattern is correct. You're using onClick in a server component which isn't allowed. If you mark the Page as client component, then you can't use the 'use server', thus, still no use.

**Reply by sofly** [reply]
@Nicky-Sa I'm using onClick in client components: `Link` and `Button` are client components. It's allowed. See [docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#client-components):
<img width="665" alt="Screenshot 2024-08-04 at 21 21 36" src="https://github.com/user-attachments/assets/63b4a92b-3a59-4deb-bef7-9d3cac7daf70">

**Reply by Nicky-Sa** [reply]
Yeah I'm aware of that, but you can only invoke a client component inside of a server component. Writing "onClick" in server is not allowed. If you try to run the code and click, you'll get an error.

**Reply by sofly** [reply]
@Nicky-Sa documentation says to us that we can only create and pass a server action to a client component, and from the client component call the server action. It's not possible to run the server action in the server component.
